### PR TITLE
Reopen user case when mobile worker is reactivated

### PR DESCRIPTION
**Pull request by NoahCarnahan**
http://manage.dimagi.com/default.asp?187418
Best reviewed commit-by-commit.
@kaapstorm cc @nickpell

### CODE REVIEW

File: corehq/apps/callcenter/utils.py [hunk 1]
```diff
@@ -116,6 +92,51 @@ def valid_element_name(name):
...
+
+    def valid_element_name(name):
+        try:
+            ElementTree.fromstring('<{}/>'.format(name))
+            return True
+        except ElementTree.ParseError:
+            return False
+
+
+    # remove any keys that aren't valid XML element names
```
Comment (dimagimon) [inline corehq/apps/callcenter/utils.py]: E303 too many blank lines (2)

**Comment by nickpell** [comment]
looks good to merge pending tests

**Comment by NoahCarnahan** [comment]
There's an unhanded edge case that I realized needs addressing. Hold off on merge until I get that straightened out.

**Comment by kaapstorm** [comment]
:+1:

**Comment by NoahCarnahan** [comment]
@kaapstorm I added a commit (15793f9) that changes the logic a bit more. It would be great if you could confirm that this approach seems sound.
- No longer submit close actions when updating already closed cases. Previously, an update to a closed case (which could happen if an archived user is modified) would also submit another close action. This made it challenging to archive the form that closes a case, which is the technique I'm using to re-open closed user cases.
- Submit two separate forms with update and close blocks if a case is updated and closed at the same time. This is also necessary for the reopening workflow.

File: corehq/apps/callcenter/utils.py [hunk 2]
```diff
@@ -52,68 +106,64 @@ def sync_user_case(commcare_user, case_type, owner_id, case=None):
     """
     with CriticalSection(['user_case_%s_for_%s' % (case_type, commcare_user._id)]):
         domain = commcare_user.project
+        fields = _get_user_case_fields(commcare_user)
+        case = case or get_case_by_domain_hq_user_id(domain.name, commcare_user._id, case_type)
+        close = commcare_user.to_be_deleted() or not commcare_user.is_active
```
Comment (kaapstorm) [inline corehq/apps/callcenter/utils.py]: Does this mean that when a user is deleted, their user case is closed, not marked as deleted? Is this correct? (It is different from the behaviour of normal cases.)

**Comment by kaapstorm** [comment]
This all looks great (assuming previous question is correct behaviour). 

Could you add a couple of tests showing what 060a87c and 15793f9 do? [CallCenterTestOpenedClosed][1] might be a useful place to add them. 

  [1]: https://github.com/dimagi/commcare-hq/blob/c2ccefafdd1861c0571dd2d505cbc731a3d5dd8f/corehq/apps/callcenter/tests/test_indicators.py#L491-L491

File: corehq/apps/callcenter/utils.py (hunk 2 above)
Comment (NoahCarnahan) [inline corehq/apps/callcenter/utils.py]: I believe that's right. When a user is deactivated, the case is indeed closed, but not deleted. It's hard to tell in the diff, but that's not actually new behavior in this PR. This particular line existed beforehand.

Comment (kaapstorm) [inline corehq/apps/callcenter/utils.py]: :+1:

**Comment by kaapstorm** [comment]
Thanks for the tests, @NoahCarnahan. I took the liberty to merge master to see whether Travis comes back green this time.

**Comment by NoahCarnahan** [comment]
Awesome, thanks Norman!

**Pull request merged**